### ISSUE TITLE
Audit never finishes on page with large SVG

### ISSUE THREAD

**Issue by mccahan** [dequelabs/axe-core#1717]
Expectation: Expected axe to finish the audit and show results when running against https://futureofhousing.aarp.org/, which contains a large SVG in-line at the top. If it's unable to complete an audit due to an issue, I would expect it to time out and give up after a reasonable amount of time, or to give some kind of warning or error message indicating where it's hung up.

The exact same page but without the large city SVG at the top (https://futureofhousing.aarp.org/?skipCity=true) completes the audit without issue.

Actual: The pane sits on "Analyzing... please wait." but never appears to finish or time out or give any error messages or other indications.

Motivation: Would expect the product to time out after a reasonable amount of time or give any kind of error indication to help troubleshoot the issue. As-is, I had to trial-and-error remove elements from the page until I found the SVG that was causing the issue.

<pre><code>
axe-core version: 3.2.2
axe-webdriver, extension or other integration version: 3.8.0 (Chrome Extension, Chrome 75.0.3770.100, Windows 10)
</code></pre>

**Comment by straker** [comment]
Thanks for reporting this. I'm able to reproduce and it seems to be stuck on the rule `duplicate-id-active`. My guess is because 3550 nodes in the SVG use the `id` of `Path`.

**Comment by sorin-davidoi** [comment]
I think this is solvable by caching the result of `isHiddenWithCSS` for the traversed nodes. From what I understand the code looks something like

```javascript
for(const node of document.querySelectorAll('[id]')) {
  const (const otherWithSameId of document.querySelectorAll(`*[id="${node.getAttribute('id')}"]`)) {
    // The following is inside isHiddenWithCSS
    const parent = otherWithSameId;
    while(parent) {
       // Some expensive computations involving getComputedStyle and getPropertyValue
      parent = parent.parentElement;
    }
  }
}
```

The problem is that `isHiddenWithCSS` can be called hundreds of time for the same element, but since we know that there have been no DOM mutations that could be avoided. The following patch makes `axe https://futureofhousing.aarp.org` finish in 1 minute 20 seconds, but I'm not sure if it's applied at the right level (or if `WeakMap`/`Map` can be used). @straker What is your take on this?

```diff
index 7c0d2f15..efe6ae9b 100644
--- a/lib/commons/dom/is-hidden-with-css.js
+++ b/lib/commons/dom/is-hidden-with-css.js
@@ -10,6 +10,24 @@
  * @return {Boolean} the element's hidden status
  */
 dom.isHiddenWithCSS = function isHiddenWithCSS(el, descendentVisibilityValue) {
+       if (!axe._cache.get('hiddenWithCSS')) {
+               axe._cache.set('hiddenWithCSS', new WeakMap());
+       }
+
+       const cache = axe._cache.get('hiddenWithCSS');
+
+       if (cache.has(el)) {
+               return cache.get(el);
+       }
+
+       const result = _isHiddenWithCSS(el, descendentVisibilityValue);
+
+       cache.set(el, result);
+
+       return result;
+}
+
+function _isHiddenWithCSS(el, descendentVisibilityValue) {
        if (el.nodeType === 9) {
                // 9 === Node.DOCUMENT
                return false;
```

**Comment by straker** [comment]
This should work great. We've cached similar results from the [is-hidden code](https://github.com/dequelabs/axe-core/blob/develop/lib/core/utils/is-hidden.js) to speed it up. We can just store the results onto the VirtualNode itself though, no need to access the `axe._cache` in this case.

**Comment by chandana7393** [comment]
Tested, working as expected.
![unique_id](https://user-images.githubusercontent.com/45285854/67858956-76bbab00-fb40-11e9-805b-487bb884b463.PNG)

**Tested Environment:**
Axe-coconut - 4.0.0.30304v
Chrome -  78.0.3904.70v
OS - Windows 10 64 bit.

**Comment by padmavemulapati** [comment]
@straker , After waiting so long time it analyzed and given results, if this is ok we can close the ticket. Please let me know if anything else required here.

**Comment by padmavemulapati** [comment]
As this slowness is tracking in other PR, which is raised with author, closing here
